### PR TITLE
Lyrics placement fixes

### PR DESCRIPTION
**Pull request by miiizen**
Resolves: #24549
Lyrics above the stave needed to be stacked from the highest voice to the lowest
Resolves:
image
The maximum verse above and below was being calculated incorrectly - we were just getting the number of verses above and below.  We also need to check if a verse is above or below the stave before adjusting other verses.

### CODE REVIEW

**Review by oktophonie — APPROVED** [review]

File: src/engraving/rendering/score/lyricslayout.cpp [hunk 1]
```diff
@@ -609,20 +609,23 @@ void LyricsLayout::setDefaultPositions(staff_idx_t staffIdx, const LyricsVersesM
     double staffHeight = ctx.dom().staff(staffIdx)->staffHeight();
     double lyricsLineHeightFactor = ctx.conf().styleD(Sid::lyricsLineHeight);
 
-    int totVersesAbove = int(lyricsVersesAbove.size()) - 1;
```
Comment (mike-spa) [inline src/engraving/rendering/score/lyricslayout.cpp:612]: As far as I can tell, the first commit can be stripped down to just this line. The error is assuming that the total number of verses coincides with the maximum verse (which is not always the case), so this should be
`int totVersesAbove = lyricsVersesAbove.empty() ? 0 : lyricsVersesAbove.rbegin()->first;`
as you rightly did. Perhaps better rename totVersesAbove as maxVerseAbove, which is more accurate.
The changes you've done to the loop below aren't changing anything, as far as I can tell, cause I'm running the loop forward with increasing y and you're running it backwards with decreasing y, which produces the same result. You could double check this by commenting out the function `checkCollisionWithStaffElements`, which lets you check the direct result of `setDefaultPositions`.

File: src/engraving/rendering/score/lyricslayout.cpp [hunk 2]
```diff
@@ -651,24 +654,28 @@ void LyricsLayout::checkCollisionsWithStaffElements(System* system, staff_idx_t
...
+    int maxVerseBelow = !lyricsVersesBelow.empty() ? lyricsVersesBelow.crbegin()->first : 0;
 
-    for (int verse = maxVerseAbove - 1; verse >= 0; --verse) {
-        SkylineLine verseSkyline = createSkylineForVerse(verse, false, lyricsVersesAbove, system);
-        double minDistance = -verseSkyline.minDistance(staffSkylineNorth);
-        if (minDistance < lyricsMinDist) {
-            double diff = lyricsMinDist - minDistance;
-            moveThisVerseAndOuterOnes(verse, 0, true, -diff, lyricsVersesAbove);
+    for (int verse = maxVerseAbove; verse >= 0; --verse) {
+        if (lyricsVersesAbove.count(verse) > 0) {
```
Comment (mike-spa) [inline src/engraving/rendering/score/lyricslayout.cpp:661]: This makes sense, good catch. I find it more readable to check-and-continue on the opposite case, cause it saves an indentation level for the main part of the function. In other words, I usually prefer
```
if (lyricsVersesAbove.count(verse) == 0) {
  continue;
}
```
but that's just coding style. Good job.

File: src/engraving/rendering/score/lyricslayout.cpp (hunk 1 above)
Comment (miiizen) [reply src/engraving/rendering/score/lyricslayout.cpp:612]: Yes, you're correct! I'll remove the unecessary changes.

**Review by mike-spa — APPROVED** [review]
Good to go for me, as long as the vtests don't bring up other problems :)

**Pull request merged**